protocols/kad/src/handler.rs

```diff
@@ -645,6 +645,8 @@ where
     ) -> Poll<
         ProtocolsHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::OutEvent, Self::Error>,
     > {
+        let had_any_substream = !self.substreams.is_empty();
+
         // We remove each element from `substreams` one by one and add them back.
         for n in (0..self.substreams.len()).rev() {
             let mut substream = self.substreams.swap_remove(n);
```

Comment:
I just realised that with the early return at the top, this loop probably needs to be adjusted for the match on `(None, Some(event), _) => {` to also do
```rust
if self.substreams.is_empty() {
    self.keep_alive = KeepAlive::Until(Instant::now() + Duration::from_secs(10));
}
```
before it does not put the substream back in the list and then `return`s from within the loop, which may otherwise result in a keep-alive previously set to `Yes` to remain so.